File: cathode-sync/src/main/java/net/simonvt/cathode/sync/scheduler/ListsTaskScheduler.java
/*
 * Copyright (C) 2015 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.sync.scheduler;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import javax.inject.Inject;
import javax.inject.Singleton;
import net.simonvt.cathode.api.enumeration.Privacy;
import net.simonvt.cathode.api.service.SyncService;
import net.simonvt.cathode.common.database.Cursors;
import net.simonvt.cathode.jobqueue.JobManager;
import net.simonvt.cathode.provider.DatabaseContract;
import net.simonvt.cathode.provider.DatabaseContract.ListItemColumns;
import net.simonvt.cathode.provider.DatabaseContract.ListsColumns;
import net.simonvt.cathode.provider.ProviderSchematic.ListItems;
import net.simonvt.cathode.provider.ProviderSchematic.Lists;
import net.simonvt.cathode.provider.helper.EpisodeDatabaseHelper;
import net.simonvt.cathode.provider.helper.ListWrapper;
import net.simonvt.cathode.provider.helper.MovieDatabaseHelper;
import net.simonvt.cathode.provider.helper.PersonDatabaseHelper;
import net.simonvt.cathode.provider.helper.SeasonDatabaseHelper;
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper;
import net.simonvt.cathode.remote.action.lists.AddEpisode;
import net.simonvt.cathode.remote.action.lists.AddMovie;
import net.simonvt.cathode.remote.action.lists.AddPerson;
import net.simonvt.cathode.remote.action.lists.AddSeason;
import net.simonvt.cathode.remote.action.lists.AddShow;
import net.simonvt.cathode.remote.action.lists.RemoveEpisode;
import net.simonvt.cathode.remote.action.lists.RemoveMovie;
import net.simonvt.cathode.remote.action.lists.RemovePerson;
import net.simonvt.cathode.remote.action.lists.RemoveSeason;
import net.simonvt.cathode.remote.action.lists.RemoveShow;
import net.simonvt.cathode.remote.sync.lists.SyncLists;
import net.simonvt.cathode.sync.trakt.UserList;

@Singleton public class ListsTaskScheduler extends BaseTaskScheduler {

  private SyncService syncService;

  private ShowDatabaseHelper showHelper;
  private SeasonDatabaseHelper seasonHelper;
  private EpisodeDatabaseHelper episodeHelper;
  private MovieDatabaseHelper movieHelper;
  private PersonDatabaseHelper personHelper;

  private UserList userList;

  @Inject
  public ListsTaskScheduler(Context context, JobManager jobManager, SyncService syncService,
      ShowDatabaseHelper showHelper, SeasonDatabaseHelper seasonHelper,
      EpisodeDatabaseHelper episodeHelper, MovieDatabaseHelper movieHelper,
      PersonDatabaseHelper personHelper, UserList userList) {
    super(context, jobManager);
    this.syncService = syncService;
    this.showHelper = showHelper;
    this.seasonHelper = seasonHelper;
    this.episodeHelper = episodeHelper;
    this.movieHelper = movieHelper;
    this.personHelper = personHelper;
    this.userList = userList;
  }

  public void createList(final String name, final String description, final Privacy privacy,
      final boolean displayNumbers, final boolean allowComments) {
    execute(new Runnable() {
      @Override public void run() {
        userList.create(name, description, privacy, displayNumbers, allowComments);
        queue(new SyncLists());
      }
    });
  }

  public void updateList(final long listId, final String name, final String description,
      final Privacy privacy, final boolean displayNumbers, final boolean allowComments) {
    execute(new Runnable() {
      @Override public void run() {
        final long traktId = ListWrapper.getTraktId(context.getContentResolver(), listId);
        userList.update(traktId, name, description, privacy, displayNumbers, allowComments);
        queue(new SyncLists());
      }
    });
  }

  public void deleteList(final long listId) {
    execute(new Runnable() {
      @Override public void run() {
        Cursor list = context.getContentResolver().query(Lists.withId(listId), new String[] {
            ListsColumns.NAME, ListsColumns.TRAKT_ID,
        }, null, null, null);
        if (list.moveToFirst()) {
          final String name = Cursors.getString(list, ListsColumns.NAME);
          final long traktId = Cursors.getLong(list, ListsColumns.TRAKT_ID);

          if (userList.delete(traktId, name)) {
            context.getContentResolver().delete(ListItems.inList(listId), null, null);
            context.getContentResolver().delete(Lists.withId(listId), null, null);
          }

          queue(new SyncLists());
        }
        list.close();
      }
    });
  }

  public void addItem(final long listId, final int itemType, final long itemId) {
    updateListItem(listId, itemType, itemId, true);
  }

  public void removeItem(final long listId, final int itemType, final long itemId) {
    updateListItem(listId, itemType, itemId, false);
  }

  private void updateListItem(final long listId, final int itemType, final long itemId,
      final boolean add) {
    execute(new Runnable() {
      @Override public void run() {
        final long listTraktId = ListWrapper.getTraktId(context.getContentResolver(), listId);
        if (add) {
          ContentValues values = new ContentValues();
          values.put(ListItemColumns.LIST_ID, listId);
          values.put(ListItemColumns.LISTED_AT, System.currentTimeMillis());
          values.put(ListItemColumns.ITEM_ID, itemId);
          values.put(ListItemColumns.ITEM_TYPE, itemType);

          context.getContentResolver().insert(ListItems.LIST_ITEMS, values);
        } else {
          context.getContentResolver()
              .delete(ListItems.inList(listId),
                  ListItemColumns.ITEM_TYPE + "=? AND " + ListItemColumns.ITEM_ID + "=?",
                  new String[] {
                      String.valueOf(itemType), String.valueOf(itemId),
                  });
        }

        switch (itemType) {
          case DatabaseContract.ItemType.SHOW: {
            final long showTraktId = showHelper.getTraktId(itemId);

            if (add) {
              queue(new AddShow(listTraktId, showTraktId));
            } else {
              queue(new RemoveShow(listTraktId, showTraktId));
            }
            break;
          }

          case DatabaseContract.ItemType.SEASON: {
            final long showId = seasonHelper.getShowId(itemId);
            final long showTraktId = showHelper.getTraktId(showId);
            final int seasonNumber = seasonHelper.getNumber(itemId);

            if (add) {
              queue(new AddSeason(listTraktId, showTraktId, seasonNumber));
            } else {
              queue(new RemoveSeason(listTraktId, showTraktId, seasonNumber));
            }
            break;
          }

          case DatabaseContract.ItemType.EPISODE: {
            final long showId = episodeHelper.getShowId(itemId);
            final long showTraktId = showHelper.getTraktId(showId);
            final int seasonNumber = episodeHelper.getSeason(itemId);
            final int episodeNumber = episodeHelper.getNumber(itemId);

            if (add) {
              queue(new AddEpisode(listTraktId, showTraktId, seasonNumber, episodeNumber));
            } else {
              queue(new RemoveEpisode(listTraktId, showTraktId, seasonNumber, episodeNumber));
            }
            break;
          }

          case DatabaseContract.ItemType.MOVIE: {
            final long movieTraktId = movieHelper.getTraktId(itemId);

            if (add) {
              queue(new AddMovie(listTraktId, movieTraktId));
            } else {
              queue(new RemoveMovie(listTraktId, movieTraktId));
            }
            break;
          }

          case DatabaseContract.ItemType.PERSON: {
            final long personTraktId = personHelper.getTraktId(itemId);

            if (add) {
              queue(new AddPerson(listTraktId, personTraktId));
            } else {
              queue(new RemovePerson(listTraktId, personTraktId));
            }
            break;
          }
        }
      }
    });
  }
}

File: cathode-sync/src/main/java/net/simonvt/cathode/sync/scheduler/ListsTaskScheduler.kt
/*
 * Copyright (C) 2015 Simon Vig Therkildsen
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.simonvt.cathode.sync.scheduler

import android.content.ContentValues
import android.content.Context
import kotlinx.coroutines.launch
import net.simonvt.cathode.actions.invokeAsync
import net.simonvt.cathode.actions.user.SyncLists
import net.simonvt.cathode.api.enumeration.ItemType
import net.simonvt.cathode.api.enumeration.Privacy
import net.simonvt.cathode.api.enumeration.SortBy
import net.simonvt.cathode.api.enumeration.SortOrientation
import net.simonvt.cathode.common.database.getLong
import net.simonvt.cathode.common.database.getString
import net.simonvt.cathode.jobqueue.JobManager
import net.simonvt.cathode.provider.DatabaseContract.ListItemColumns
import net.simonvt.cathode.provider.DatabaseContract.ListsColumns
import net.simonvt.cathode.provider.ProviderSchematic.ListItems
import net.simonvt.cathode.provider.ProviderSchematic.Lists
import net.simonvt.cathode.provider.helper.EpisodeDatabaseHelper
import net.simonvt.cathode.provider.helper.ListDatabaseHelper
import net.simonvt.cathode.provider.helper.MovieDatabaseHelper
import net.simonvt.cathode.provider.helper.PersonDatabaseHelper
import net.simonvt.cathode.provider.helper.SeasonDatabaseHelper
import net.simonvt.cathode.provider.helper.ShowDatabaseHelper
import net.simonvt.cathode.provider.query
import net.simonvt.cathode.remote.action.lists.AddEpisode
import net.simonvt.cathode.remote.action.lists.AddMovie
import net.simonvt.cathode.remote.action.lists.AddPerson
import net.simonvt.cathode.remote.action.lists.AddSeason
import net.simonvt.cathode.remote.action.lists.AddShow
import net.simonvt.cathode.remote.action.lists.RemoveEpisode
import net.simonvt.cathode.remote.action.lists.RemoveMovie
import net.simonvt.cathode.remote.action.lists.RemovePerson
import net.simonvt.cathode.remote.action.lists.RemoveSeason
import net.simonvt.cathode.remote.action.lists.RemoveShow
import net.simonvt.cathode.settings.TraktLinkSettings
import net.simonvt.cathode.sync.trakt.UserList
import javax.inject.Inject
import javax.inject.Singleton

@Singleton
class ListsTaskScheduler @Inject constructor(
  context: Context,
  jobManager: JobManager,
  private val showHelper: ShowDatabaseHelper,
  private val seasonHelper: SeasonDatabaseHelper,
  private val episodeHelper: EpisodeDatabaseHelper,
  private val movieHelper: MovieDatabaseHelper,
  private val personHelper: PersonDatabaseHelper,
  private val listHelper: ListDatabaseHelper,
  private val userList: UserList,
  private val syncLists: SyncLists
) : BaseTaskScheduler(context, jobManager) {

  fun createList(
    name: String,
    description: String,
    privacy: Privacy,
    displayNumbers: Boolean,
    allowComments: Boolean,
    sortBy: SortBy,
    sortOrientation: SortOrientation
  ) {
    scope.launch {
      if (TraktLinkSettings.isLinked(context)) {
        userList.create(
          name,
          description,
          privacy,
          displayNumbers,
          allowComments,
          sortBy,
          sortOrientation
        )
        syncLists.invokeAsync(SyncLists.Params())
      } else {
        val values = ContentValues()
        values.put(ListsColumns.NAME, name)
        values.put(ListsColumns.DESCRIPTION, description)
        values.put(ListsColumns.PRIVACY, Privacy.PRIVATE.toString())
        values.put(ListsColumns.DISPLAY_NUMBERS, false)
        values.put(ListsColumns.ALLOW_COMMENTS, true)
        values.put(ListsColumns.SORT_BY, sortBy.toString())
        values.put(ListsColumns.SORT_ORIENTATION, sortOrientation.toString())
        context.contentResolver.insert(Lists.LISTS, values)
      }
    }
  }

  fun updateList(
    listId: Long,
    name: String,
    description: String,
    privacy: Privacy,
    displayNumbers: Boolean,
    allowComments: Boolean,
    sortBy: SortBy,
    sortOrientation: SortOrientation
  ) {
    scope.launch {
      if (TraktLinkSettings.isLinked(context)) {
        val traktId = listHelper.getTraktId(listId)
        userList.update(
          traktId,
          name,
          description,
          privacy,
          displayNumbers,
          allowComments,
          sortBy,
          sortOrientation
        )
        syncLists.invokeAsync(SyncLists.Params())
      } else {
        val values = ContentValues()
        values.put(ListsColumns.NAME, name)
        values.put(ListsColumns.DESCRIPTION, description)
        values.put(ListsColumns.PRIVACY, Privacy.PRIVATE.toString())
        values.put(ListsColumns.DISPLAY_NUMBERS, false)
        values.put(ListsColumns.ALLOW_COMMENTS, true)
        values.put(ListsColumns.SORT_BY, sortBy.toString())
        values.put(ListsColumns.SORT_ORIENTATION, sortOrientation.toString())
        context.contentResolver.update(Lists.withId(listId), values, null, null)
      }
    }
  }

  fun deleteList(listId: Long) {
    scope.launch {
      val list = context.contentResolver.query(
        Lists.withId(listId),
        arrayOf(ListsColumns.NAME, ListsColumns.TRAKT_ID)
      )
      if (list.moveToFirst()) {
        val name = list.getString(ListsColumns.NAME)

        if (TraktLinkSettings.isLinked(context)) {
          val traktId = list.getLong(ListsColumns.TRAKT_ID)

          if (userList.delete(traktId, name)) {
            context.contentResolver.delete(ListItems.inList(listId), null, null)
            context.contentResolver.delete(Lists.withId(listId), null, null)
          }

          syncLists.invokeAsync(SyncLists.Params())
        } else {
          context.contentResolver.delete(ListItems.inList(listId), null, null)
          context.contentResolver.delete(Lists.withId(listId), null, null)
        }
      }
      list.close()
    }
  }

  fun addItem(listId: Long, itemType: ItemType, itemId: Long) {
    updateListItem(listId, itemType, itemId, true)
  }

  fun removeItem(listId: Long, itemType: ItemType, itemId: Long) {
    updateListItem(listId, itemType, itemId, false)
  }

  private fun updateListItem(listId: Long, itemType: ItemType, itemId: Long, add: Boolean) {
    scope.launch {
      val listTraktId = listHelper.getTraktId(listId)
      if (add) {
        val values = ContentValues()
        values.put(ListItemColumns.LIST_ID, listId)
        values.put(ListItemColumns.LISTED_AT, System.currentTimeMillis())
        values.put(ListItemColumns.ITEM_ID, itemId)
        values.put(ListItemColumns.ITEM_TYPE, itemType.toString())

        context.contentResolver.insert(ListItems.LIST_ITEMS, values)
      } else {
        context.contentResolver
          .delete(
            ListItems.inList(listId),
            ListItemColumns.ITEM_TYPE + "=? AND " + ListItemColumns.ITEM_ID + "=?",
            arrayOf(itemType.toString(), itemId.toString())
          )
      }

      if (TraktLinkSettings.isLinked(context)) {
        when (itemType) {
          ItemType.SHOW -> {
            val showTraktId = showHelper.getTraktId(itemId)

            if (add) {
              queue(AddShow(listTraktId, showTraktId))
            } else {
              queue(RemoveShow(listTraktId, showTraktId))
            }
          }

          ItemType.SEASON -> {
            val showId = seasonHelper.getShowId(itemId)
            val showTraktId = showHelper.getTraktId(showId)
            val seasonNumber = seasonHelper.getNumber(itemId)

            if (add) {
              queue(AddSeason(listTraktId, showTraktId, seasonNumber))
            } else {
              queue(RemoveSeason(listTraktId, showTraktId, seasonNumber))
            }
          }

          ItemType.EPISODE -> {
            val showId = episodeHelper.getShowId(itemId)
            val showTraktId = showHelper.getTraktId(showId)
            val seasonNumber = episodeHelper.getSeason(itemId)
            val episodeNumber = episodeHelper.getNumber(itemId)

            if (add) {
              queue(AddEpisode(listTraktId, showTraktId, seasonNumber, episodeNumber))
            } else {
              queue(RemoveEpisode(listTraktId, showTraktId, seasonNumber, episodeNumber))
            }
          }

          ItemType.MOVIE -> {
            val movieTraktId = movieHelper.getTraktId(itemId)

            if (add) {
              queue(AddMovie(listTraktId, movieTraktId))
            } else {
              queue(RemoveMovie(listTraktId, movieTraktId))
            }
          }

          ItemType.PERSON -> {
            val personTraktId = personHelper.getTraktId(itemId)

            if (add) {
              queue(AddPerson(listTraktId, personTraktId))
            } else {
              queue(RemovePerson(listTraktId, personTraktId))
            }
          }

          else -> throw RuntimeException("Unknown item type: $itemType")
        }
      }
    }
  }
}
